include/swift/AST/ASTContext.h

```diff
@@ -869,6 +869,27 @@ class ASTContext final {
   /// This guarantees that resulted \p names doesn't have duplicated names.
   void getVisibleTopLevelModuleNames(SmallVectorImpl<Identifier> &names) const;
 
+  struct OverrideSignatureKey {
+    StringRef baseMethodSigString;
+    StringRef derivedClassSigString;
+    Type superclassTy;
+
+    OverrideSignatureKey(StringRef baseMethodSig, StringRef derivedClassSig,
+                         Type derivedClassSuperclassType)
+        : baseMethodSigString(baseMethodSig),
+          derivedClassSigString(derivedClassSig),
+          superclassTy(derivedClassSuperclassType) {}
+
+    bool isEqual(const OverrideSignatureKey other) {
+      return baseMethodSigString == other.baseMethodSigString &&
+             derivedClassSigString == other.derivedClassSigString &&
+             superclassTy.getString() == other.superclassTy.getString();
+    }
+  };
+
+  std::vector<std::pair<OverrideSignatureKey, GenericSignature *>>
```

Comment:
Please use a DenseMap (and I know there's a bunch of boilerplate to define equality and hashing for it...)